text/0006-monolothic-packaging.md

```diff
@@ -0,0 +1,343 @@
+  "dependencies": {
+    "aws-cdk-lib": "2.89.0",
+    "constructs": "1.77.0",
+    "your-awesome-lib": "^1.0.0"
+  }
+}
+```
+
+In JavaScript code, import statements that use the CDK will now look like this:
+
+```ts
+import { aws_s3, aws_dynamodb } from 'aws-cdk-lib';
+import { Construct } from 'constructs';
+
+// ...
+export class MyConstruct extends Construct {
+  constructor(scope: Construct, id: string) {
+    super(scope, id);
+
+    new aws_s3.Bucket(this, ...);
+    new aws_dynamodb.Table(this, ...);
+  }
+}
+```
+
+> TODO: other languages
+
+# Design Summary
+
+The general approach proposed by this RFC is to ship the core types and the
+entire AWS Construct Library as a single module for all languages.
+
+This will dramatically simplify how users declare their compatibility with the
+AWS CDK, and I would argue it is also more aligned with our user's mental model (see [Rational](#rationale-and-alternatives) beflow).
+
+# Detailed Design
+
+## Prototype
+
+See [monocdk-experiment](https://www.npmjs.com/package/monocdk-experiment).
+
+[PR](https://github.com/awslabs/aws-delivlib/pull/245) with the migration of aws-delivlib to monocdk.
+
+## cx-api
+
+The cx-api module is used to coordinate the protocol between the CDK apps and the CLI. Today, both the CLI and the framework are dynamically linked against this module (it is defined in `dependencies`). Once we ship the CDK as a single monolithic module, we will need to decide how to coordinate the protocol.
+
+The current thinking is to *statically link* the protocol definitions between the CLI and the framework. This means that at build time we will create a copy of the protocol definitions (the `.ts` files) and use it both in the CLI and the framework.
+
+## TODO
+
+- [ ] We lose per-module analytics which means we will to move to report
+  analytics at the construct level.
+- [ ] We jsii to support the concept of "submodules". I believe this is easier
+  than arbitrary namespaces which we decided not to support in jsii. One thing to consider for example, is where do submodule README files go.
+- [ ] `constructs`
+- [ ] `@aws-cdk/assert` library. Do we have to do the assert library rewrite as a jsii module (we want to do that anyway!).
+- [ ] Rewrite all example code (see [cdk-rewrite-mono-import](https://github.com/rix0rrr/cdk-rewrite-mono-imports)).
+- [ ] Reference documentation needs to also support submodules/namespaces and use the submodule's README file.
+
+# Drawbacks
+
+- [ ] In JavaScript/TypeScript code will always be fully qualified
+  (`aws_s3.Bucket` instead of `Bucket`) due to limitations of the `import` tool.
+- [ ] The size of the single module will be large (~50MiB) which can create issues of using the CDK in non-standard environments such as from a Lambda function.
+- [ ] Breaking change
+
+# Rationale and Alternatives
+
+I believe our users think of the AWS CDK as a "standard library" (or a
+"framework") and not as another library that they depend on to build their
```

Comment:
Many frameworks are not monolithic. Angular and React on the web, for instance, are not, and not solely due to size concerns. There is still a distinction between what is considered "essential" and what isn't. How is this a strong rationale for monolithic publishing?